napari/config.py

```diff
@@ -0,0 +1,877 @@
+    turns these into config variables of the form ``{"foo": {"bar-baz": 123}}``
+    It transforms the key and value in the following way:
+
+    -  Lower-cases the key text
+    -  Treats ``__`` (double-underscore) as nested access
+    -  Calls ``ast.literal_eval`` on the value
+    """
+    if env is None:
+        env = os.environ
+    d = {}
+    for name, value in env.items():
+        if name.startswith("NAPARI_"):
+            varname = name[7:].lower().replace("__", ".")
+            try:
+                d[varname] = ast.literal_eval(value)
+            except (SyntaxError, ValueError):
+                d[varname] = value
+
+    result: dict = {}
+    set(d, config=result, clean=True)
+
+    return result
+
+
+def ensure_file(source: str, destination: Optional[str] = None, comment=True):
+    """Copy file to default location if it does not already exist.
+
+    This tries to move a default configuration file to a default location if if
+    does not already exist.  It also comments out that file by default.
+
+    This is to be used by downstream modules that may have default
+    configuration files that they wish to include in the default configuration
+    path.
+
+    Parameters
+    ----------
+    source : string, filename
+        Source configuration file, typically within a source directory.
+    destination : string, directory
+        Destination directory. Configurable by ``NAPARI_CONFIG`` environment
+        variable, falling back to ~/.config/napari.
+    comment : bool, True by default
+        Whether or not to comment out the config file when copying.
+    """
+    if destination is None:
+        destination = PATH
+
+    # destination is a file and already exists, never overwrite
+    if os.path.isfile(destination):
+        return
+
+    # If destination is not an existing file, interpret as a directory,
+    # use the source basename as the filename
+    directory = destination
+    destination = os.path.join(directory, os.path.basename(source))
+
+    try:
+        if not os.path.exists(destination):
+            os.makedirs(directory, exist_ok=True)
+
+            # Atomically create destination.  Parallel testing discovered
+            # a race condition where a process can be busy creating the
+            # destination while another process reads an empty config file.
+            tmp = "%s.tmp.%d" % (destination, os.getpid())
+            with open(source) as f:
+                lines = list(f)
+
+            if comment:
+                lines = [
+                    "# " + line
+                    if line.strip() and not line.startswith("#")
+                    else line
+                    for line in lines
+                ]
+
+            with open(tmp, "w") as f:
+                f.write("".join(lines))
+
+            try:
+                os.rename(tmp, destination)
+            except OSError:
+                os.remove(tmp)
+    except (IOError, OSError):
+        pass
+
+
+class set:
```

Comment:
Oh. God. :joy:

Reply:
took a quick look at implementing with `@contextmanager`, but I think the additional `_assign` method on the class makes it uglier to implement this as a function